src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/SNIStreams.ValueTask.cs

```diff
@@ -0,0 +1,122 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Threading;
+using System.Threading.Tasks;
+using System;
+
+namespace Microsoft.Data.SqlClient.SNI
+{
+    internal sealed partial class SNISslStream
+    {
+        public override Task<int> ReadAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            ValueTask<int> valueTask = ReadAsync(new Memory<byte>(buffer, offset, count), cancellationToken);
+            if (valueTask.IsCompletedSuccessfully)
```

Comment:
This seems redundant - ValueTask.AsTask already knows to efficiently return a completed Task.

Though I'd personally just call the corresponding Task-returning method on the base class directly, and apply the semaphore check here. At least theoretically, the Task-returning method could be doing something slightly different since it knows it's returning a Task (e.g. no need to deal with IValueTaskSource etc.).

Reply:
So you think it would be better to just do `return ReadAsync(new Memory<byte>(buffer, offset, count), cancellationToken).AsTask();` ?

Reply:
Yeah, it's basically the same. Take a look at the source for AsTask.